### PR TITLE
Expand the transit backend features

### PR DESCRIPTION
**Pull request by EraYaN**
Right now this PR adds the following Endpoints:
- Rewrap
- List

Planning to add:
- Maybe some from #12 

It also adds all the new Key types supported by the latest version of the transit backend.

### CODE REVIEW

File: src/VaultSharp/V1/SecretsEngines/Transit/ITransitSecretsEngine.cs [hunk 1]
```diff
@@ -8,6 +8,19 @@ namespace VaultSharp.V1.SecretsEngines.Transit
...
+        /// Only the key names are returned (not the actual keys themselves).
+        /// </summary>
+        /// <param name="mountPoint"><para>[optional]</para>
+        /// The mount point for the Transit backend. Defaults to <see cref="SecretsEngineDefaultPaths.Transit" />
+        /// Provide a value only if you have customized the mount point.
+        /// </param>
+        /// <returns>
+        /// The key names.
+        /// </returns>
+        Task<Secret<ListResponse>> ListAsync(string mountPoint = SecretsEngineDefaultPaths.Transit);
```
Comment (Marusyk) [inline src/VaultSharp/V1/SecretsEngines/Transit/ITransitSecretsEngine.cs]: Please add `wrapTimeToLive` parameter. 
More info about [wrapTimeToLiv](https://www.vaultproject.io/docs/concepts/response-wrapping/#response-wrapping-token-creation)

File: src/VaultSharp/Core/Polymath.cs [hunk 2]
```diff
@@ -119,45 +119,11 @@ public async Task MakeVaultApiRequest(string resourcePath, HttpMethod httpMethod
...
-                        break;
-
-                    case "HEAD":
-
-                        httpRequestMessage = new HttpRequestMessage(HttpMethod.Head, requestUri);
-                        break;
-
-                    default:
-                        throw new NotSupportedException("The Http Method is not supported: " + httpMethod);
+                    httpRequestMessage.Content = requestContent;
```
Comment (Marusyk) [inline src/VaultSharp/Core/Polymath.cs]: Your refactored code is great. But before, we threw `NotSupportedException` when the HTTP method is unsupported (like OPTIONS, PATCH, etc). I think we should leave that behavior to support backward compatibility. Could you please make small changes to handle it?

File: src/VaultSharp/V1/SecretsEngines/Transit/TransitSecretsEngineProvider.cs [hunk 3]
```diff
@@ -14,6 +14,11 @@ public TransitSecretsEngineProvider(Polymath polymath)
             _polymath = polymath;
         }
 
+        public async Task<Secret<ListResponse>> ListAsync(string mountPoint = SecretsEngineDefaultPaths.Transit)
+        {
+            return await _polymath.MakeVaultApiRequest<Secret<ListResponse>>("v1/" + mountPoint.Trim('/') + "/keys", new HttpMethod("LIST")).ConfigureAwait(_polymath.VaultClientSettings.ContinueAsyncTasksOnCapturedContext);
```
Comment (Marusyk) [inline src/VaultSharp/V1/SecretsEngines/Transit/TransitSecretsEngineProvider.cs]: Be careful with the HTTP method LIST, it should be properly handled here
https://github.com/VaultSharp/VaultSharp/pull/13/files#diff-9c1995235b4bb2ad85b7259205024c8bR122

File: test/VaultSharp.Samples/Program.cs [hunk 4]
```diff
@@ -183,66 +183,98 @@ private static void RunTransitSamples()
...
 
-            var keyName = "test_key";
+            var listResponse = _authenticatedVaultClient.V1.Secrets.Transit.ListAsync().Result;
+            var keyNames = listResponse.Data.Keys;
 
-            // .\vault.exe secrets enable transit
-            // .\vault.exe write -f transit/keys/test_key
+            if (keyNames.Length > 0)
+            {
+                var keyName = keyNames[0];
```
Comment (Marusyk) [inline test/VaultSharp.Samples/Program.cs]: Please remove spaces at the end of line

File: test/VaultSharp.Samples/Program.cs [hunk 5]
```diff
@@ -183,66 +183,98 @@ private static void RunTransitSamples()
...
+                var encryptionResponse = _authenticatedVaultClient.V1.Secrets.Transit.EncryptAsync(keyName, encryptOptions).Result;
+                var cipherText = encryptionResponse.Data.CipherText;
 
-            encryptOptions = new EncryptRequestOptions
-            {
-                BatchedEncryptionItems = new List<EncryptionItem>
+                encryptOptions = new EncryptRequestOptions
+                {
+                    BatchedEncryptionItems = new List<EncryptionItem>
                 {
```
Comment (Marusyk) [inline test/VaultSharp.Samples/Program.cs]: Could you please move braces to the right?

File: test/VaultSharp.Samples/Program.cs [hunk 6]
```diff
@@ -183,66 +183,98 @@ private static void RunTransitSamples()
...
+                var decryptionResponse = _authenticatedVaultClient.V1.Secrets.Transit.DecryptAsync(keyName, decryptOptions).Result;
+                var gotPlainText = decryptionResponse.Data.Base64EncodedPlainText;
 
-            decryptOptions = new DecryptRequestOptions
-            {
-                BatchedDecryptionItems = new List<DecryptionItem>
+                decryptOptions = new DecryptRequestOptions
+                {
+                    BatchedDecryptionItems = new List<DecryptionItem>
                 {
```
Comment (Marusyk) [inline test/VaultSharp.Samples/Program.cs]: Could you please move braces to the right?

File: test/VaultSharp.Samples/Program.cs [hunk 7]
```diff
@@ -183,66 +183,98 @@ private static void RunTransitSamples()
...
+                    Nonce = nonce
+                };
+
+                var rewrapResponse = _authenticatedVaultClient.V1.Secrets.Transit.RewrapAsync(keyName, rewrapOptions).Result;
+                cipherText = rewrapResponse.Data.CipherText;
+
+                rewrapOptions = new RewrapRequestOptions
+                {
+                    BatchedRewrapItems = new List<RewrapItem>
+                {
```
Comment (Marusyk) [inline test/VaultSharp.Samples/Program.cs]: Could you please move braces to the right?

File: src/VaultSharp/V1/SecretsEngines/Transit/TransitSecretsEngineProvider.cs (hunk 3 above)
Comment (EraYaN) [reply src/VaultSharp/V1/SecretsEngines/Transit/TransitSecretsEngineProvider.cs]: It seemed to me LIST is essentially a GET with a non standard name.

File: src/VaultSharp/Core/Polymath.cs (hunk 2 above)
Comment (EraYaN) [reply src/VaultSharp/Core/Polymath.cs]: Do you want a check on a whitelist of HTTP methods Vault supports? Cause the original implementation had some code duplication.

https://github.com/hashicorp/vault/blob/master/http/cors.go#L13-L18

The list seems to be

- OPTIONS
- POST
- PUT
- DELETE
- GET
- LIST

File: test/VaultSharp.Samples/Program.cs (hunk 5 above)
Comment (EraYaN) [reply test/VaultSharp.Samples/Program.cs]: What do you mean?

```
            encryptOptions = new EncryptRequestOptions
            {
                BatchedEncryptionItems = new List<EncryptionItem>                    
```  
to
```
            encryptOptions = new EncryptRequestOptions
                {
                BatchedEncryptionItems = new List<EncryptionItem>                    
```  

Like that?

File: src/VaultSharp/Core/Polymath.cs (hunk 2 above)
Comment (EraYaN) [reply src/VaultSharp/Core/Polymath.cs]: I added it back in, even though the server would have generated a server side exception. This is a bit more maintenance sensitive if they ever change that list.

File: test/VaultSharp.Samples/Program.cs (hunk 5 above)
Comment (Marusyk) [reply test/VaultSharp.Samples/Program.cs]: No, I tell about line 220

Comment (Marusyk) [reply test/VaultSharp.Samples/Program.cs]: Screenshot_2

Comment (EraYaN) [reply test/VaultSharp.Samples/Program.cs]: Aah applied.

**Pull request merged**